recipe-server/client/actions/tests/test_show-heartbeat.js

```diff
@@ -82,112 +56,300 @@ describe('ShowHeartbeatAction', () => {
+        await firstAction.execute();
+
+        // first time should run fine
+        expect(normandy.showHeartbeat).toHaveBeenCalled();
+
+        // clear the history of calls from the testing
+        normandy.showHeartbeat.calls.reset();
+
+        const secondRecipe = recipeFactory({ ...nagConfig });
+        const secondAction = new ShowHeartbeatAction(normandy, secondRecipe);
+        await secondAction.execute();
+        // second time should NOT run
+        expect(normandy.showHeartbeat).not.toHaveBeenCalled();
+      });
+
+      it('should set the last-shown date', async () => {
+        const action = new ShowHeartbeatAction(normandy, recipeFactory({ ...nagConfig }));
+
+        spyOn(Date, 'now').and.returnValue(10);
+
+        expect(normandy.mock.storage.data.lastShown).toBeUndefined();
+        await action.execute();
+        expect(normandy.mock.storage.data.lastShown).toEqual('10');
+      });
+
+      it('should show if a user has not interacted with it', async () => {
+        const testRecipe = recipeFactory({ ...nagConfig });
+        const testAction = new ShowHeartbeatAction(normandy, testRecipe);
+        await testAction.execute();
+        expect(normandy.showHeartbeat).toHaveBeenCalled();
+      });
+
+      it('should not show if the user has interacted with it', async() => {
+        const testAction = new ShowHeartbeatAction(normandy, recipeFactory({ ...nagConfig }));
+        expect(normandy.showHeartbeat).not.toHaveBeenCalled();
+
+        // fake the 'last interacted' value
+        testAction.updateLastInteraction();
+
+        await testAction.execute();
+
+        expect(normandy.showHeartbeat).not.toHaveBeenCalled();
+      });
     });
-    const action = new ShowHeartbeatAction(normandy, recipe);
 
-    normandy.mock.client.version = '42.0.1';
-    normandy.mock.client.channel = 'nightly';
-    normandy.mock.client.isDefaultBrowser = true;
-    normandy.mock.client.searchEngine = 'shady-tims';
-    normandy.mock.client.syncSetup = true;
 
-    await action.execute();
-    const postAnswerUrl = normandy.showHeartbeat.calls.argsFor(0)[0].postAnswerUrl;
-    const params = new URL(postAnswerUrl).searchParams;
-    expect(params.get('utm_source')).toEqual('firefox');
-    expect(params.get('utm_medium')).toEqual('show-heartbeat');
-    expect(params.get('utm_campaign')).toEqual('Thisisatestmessage%21%21');
+    describe('`xdays`', () => {
+      afterEach(() => {
+        normandy.showHeartbeat.calls.reset();
+        delete normandy.mock.storage.data.lastShown;
+      });
+
+      const ONE_DAY = (1000 * 3600 * 24); // in milliseconds
+      const repeatEvery = 7; // days
+      const xdaysConfig = {
+        arguments: {
+          repeatOption: 'xdays',
+          repeatEvery,
+        },
+      };
+
+      it('should not show if less than `repeatEvery` days has elapsed', async () => {
+        const action = new ShowHeartbeatAction(normandy, recipeFactory({ ...xdaysConfig }));
+
+        let idx = 0;
```

Comment:
Old habit - using `for (let i=0; ...` creates a new variable on each loop, defining it beforehand will only use the one instance. I'll change it.